test/client-mock.js

```diff
@@ -0,0 +1,650 @@
+'use strict'
+
+const { test } = require('tap')
+const { createServer } = require('http')
+const { promisify } = require('util')
+const { Client, MockClient, request } = require('..')
+const { mockDispatch } = require('../lib/client-mock')
+const { kUrl } = require('../lib/core/symbols')
+
+async function getResponse (body) {
+  const buffers = []
+  for await (const data of body) {
+    buffers.push(data)
+  }
+
+  return Buffer.concat(buffers).toString('utf8')
+}
+
+test('mockDispatch - should handle a single interceptor', (t) => {
+  t.plan(1)
+
+  const baseUrl = 'http://localhost:9999'
+  const mockClient = new MockClient(baseUrl)
+  t.tearDown(mockClient.close.bind(mockClient))
+
+  mockClient.intercept({
+    path: '/foo',
+    method: 'GET'
+  }).reply(200, { foo: 'bar' })
+
+  this[kUrl] = new URL('http://localhost:9999')
+  mockDispatch.bind(this)({
+    path: '/foo',
+    method: 'GET'
+  }, {
+    onHeaders: (_statusCode, _headers, resume) => resume(),
+    onData: () => {},
+    onComplete: () => {}
+  })
+  t.ok(1)
+})
+
+test('ClientMock - basic intercept', (t) => {
+  t.plan(4)
+
+  const server = createServer((req, res) => {
+    res.setHeader('content-type', 'text/plain')
+    res.end('should not be called')
+    t.fail('should not be called')
+    t.end()
+  })
+  t.tearDown(server.close.bind(server))
+
+  server.listen(0, async () => {
+    const baseUrl = `http://localhost:${server.address().port}`
+
+    const client = new Client(baseUrl)
+    t.tearDown(client.close.bind(client))
+
+    const mockClient = new MockClient(baseUrl)
+    t.tearDown(mockClient.close.bind(mockClient))
+    mockClient.intercept({
+      path: '/foo?hello=there&see=ya',
+      method: 'POST',
+      body: 'form1=data1&form2=data2'
+    }).reply(200, { foo: 'bar' }, {
+      headers: {
+        'content-type': 'application/json'
+      },
+      trailers: { 'Content-MD5': 'test' }
+    })
+
+    try {
+      const { statusCode, headers, trailers, body } = await client.request({
+        path: '/foo?hello=there&see=ya',
+        method: 'POST',
+        body: 'form1=data1&form2=data2'
+      })
+      t.strictEqual(statusCode, 200)
+      t.strictEqual(headers['content-type'], 'application/json')
+      t.deepEqual(trailers, { 'content-md5': 'test' })
+
+      const jsonResponse = JSON.parse(await getResponse(body))
+      t.deepEqual(jsonResponse, {
+        foo: 'bar'
+      })
+    } catch (err) {
+      t.fail(err)
+    }
+  })
+})
+
+test('ClientMock - basic intercept with request', (t) => {
+  t.plan(2)
+
+  const server = createServer((req, res) => {
+    res.setHeader('content-type', 'text/plain')
+    res.end('should not be called')
+    t.fail('should not be called')
+    t.end()
+  })
+  t.tearDown(server.close.bind(server))
+
+  server.listen(0, async () => {
```

Comment:
please use promisify on listen, let's not pass async callbacks to functions.

Reply:
sure thing, on it now!

Reply:
fixed, my fault - should've sorted that straight away!